Model structure visualization using dot

vzhong opened the pull request:
Hey Francois,

I wrote a small utility to visualize the network structure using `pydot`. For example, the follow code:

```python
ent_lookup = Sequential()
ent_lookup.add(Embedding(10, 2))
ent_lookup.add(Flatten())

rel_lookup = Sequential()
rel_lookup.add(Embedding(20, 2))
rel_lookup.add(Flatten())

word_sequence = Sequential()
word_sequence.add(Embedding(10, 5))
word_sequence.add(GRU(5, 2))

model = Sequential()
model.add(Merge([word_sequence, ent_lookup, rel_lookup], mode='concat'))
model.add(Activation('relu'))
model.add(Dense(6, 2))
model.add(Activation('softmax'))

g = Grapher()
g.plot(model, 'mymodel.png')
```

generates something like:

![mymodel](https://cloud.githubusercontent.com/assets/1855260/7874389/b008cc76-0563-11e5-8be4-65c5cddcf7a5.png)

Let me know if you want to merge this. If so, I can polish up the code, add comments and docs etc...

fchollet commented on keras/utils/dot_utils.py:
```diff
@@ -0,0 +1,41 @@
+    def get_name(self, model):
+        if hasattr(model, 'name'):
+            return model.name
+        cls = model.__class__.__name__
```
`cls` is a reserved keyword, it would be better to use a different variable name.

fchollet commented:
This looks great. Having a paragraph about it in the docs would be neat, certainly. I'll merge this when you tell me it's ready : )

vzhong commented:
Cool! Good catch on the reserved word :+1: I've changed the reserved word and added some documentation. I wasn't sure where to put the documentation because we don't have one for `utils`, so I've gone ahead and made one. I put this class inside "Visualization Utilities", which itself is under "Utils". Please feel free to change this.

lukedeo commented:
Should we add some more explicit support for naming layers? I see thats where we get the names on the graph here. Also might be useful for logging...

vzhong commented:
Actually the way this works right now, it assigns generic names for all instance based on the rule `<class_name><number_of_class_intances_seen>`. If the model has a `name` attribute, then that name will be used in the graph.

I agree that supporting explicit names may be useful. It would probably be useful for logging, visualization, and debugging (eg. we can annotate the Theano graph with names).

fchollet commented:
Indeed, it could be nice to have an optional `name` argument that could be passed to any layer, and would then be reused in layer visualizations / debugging. I'll look at that very soon...
